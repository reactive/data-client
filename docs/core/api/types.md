---
title: TypeScript Types
---

## Manager

```typescript
interface Manager<Actions = ActionTypes> {
  middleware: Middleware<Actions>;
  cleanup(): void;
  init?: (state: State<any>) => void;
}
```

```typescript
type Middleware<Actions = any> = <C extends Controller<Actions>>(
  controller: C,
) => (next: C['dispatch']) => C['dispatch'];
```

[More](./Manager) about manager.

## NetworkError

```typescript
interface NetworkError extends Error {
  status: number;
  response?: Response;
}
```

## UnknownError

This is a catch-all for errors thrown in fetch functions. It is recommended
to try to conform to the `NetworkError` interface above

```typescript
type UnknownError = Error & { status?: unknown; response?: unknown };
```

## State

```typescript
interface State<T> {
  readonly entities: {
    readonly [entityKey: string]: { readonly [pk: string]: T } | undefined;
  };
  readonly indexes: NormalizedIndex;
  readonly results: { readonly [key: string]: unknown | PK[] | PK | undefined };
  readonly meta: {
    readonly [key: string]: {
      readonly date: number;
      readonly error?: ErrorTypes;
      readonly expiresAt: number;
      readonly prevExpiresAt?: number;
      readonly invalidated?: boolean;
      readonly errorPolicy?: 'hard' | 'soft' | undefined;
    };
  };
  readonly entitiesMeta: {
    readonly [entityKey: string]: {
      readonly [pk: string]: {
        readonly date: number;
        readonly expiresAt: number;
        readonly fetchedAt: number;
      };
    };
  };
  readonly optimistic: (
    | SetAction
    | OptimisticAction
  )[];
  readonly lastReset: number;
}
```
