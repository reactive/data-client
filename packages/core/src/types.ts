import { NormalizedIndex } from '@data-client/normalizr';
import type {
  UpdateFunction,
  AbstractInstanceType,
} from '@data-client/normalizr';
import type { ErrorTypes } from '@data-client/normalizr';

import type {
  ActionTypes,
  SetResponseAction,
  OptimisticAction,
} from './actions.js';
import type { Dispatch, Middleware, MiddlewareAPI } from './middlewareTypes.js';

export * from './actions.js';

export type { AbstractInstanceType, UpdateFunction };

export type PK = string;

/** Normalized state for Reactive Data Client
 *
 * @see https://dataclient.io/docs/concepts/normalization
 */
export interface State<T> {
  readonly entities: {
    readonly [entityKey: string]: { readonly [pk: string]: T } | undefined;
  };
  readonly endpoints: {
    readonly [key: string]: unknown | PK[] | PK | undefined;
  };
  readonly indexes: NormalizedIndex;
  readonly meta: {
    readonly [key: string]: {
      readonly date: number;
      readonly fetchedAt: number;
      readonly expiresAt: number;
      readonly prevExpiresAt?: number;
      readonly error?: ErrorTypes;
      readonly invalidated?: boolean;
      readonly errorPolicy?: 'hard' | 'soft' | undefined;
    };
  };
  readonly entitiesMeta: {
    readonly [entityKey: string]: {
      readonly [pk: string]: {
        readonly fetchedAt: number;
        readonly date: number;
        readonly expiresAt: number;
      };
    };
  };
  readonly optimistic: (SetResponseAction | OptimisticAction)[];
  readonly lastReset: number;
}

/** Singletons that handle global side-effects
 *
 * Kind of like useEffect() for the central data store
 *
 * @see https://dataclient.io/docs/api/Manager
 */
export interface Manager<Actions = ActionTypes> {
  /** @see https://dataclient.io/docs/api/Manager#getmiddleware */
  getMiddleware?(): Middleware<Actions>;
  /** @see https://dataclient.io/docs/api/Manager#middleware */
  middleware?: Middleware<Actions>;
  /** @see https://dataclient.io/docs/api/Manager#cleanup */
  cleanup(): void;
  /** @see https://dataclient.io/docs/api/Manager#init */
  init?: (state: State<any>) => void;
}

export type { Dispatch, Middleware, MiddlewareAPI };
