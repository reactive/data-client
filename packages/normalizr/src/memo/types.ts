import WeakDependencyMap from './WeakDependencyMap.js';
import { EntityInterface } from '../interface.js';
import { EntityPath } from '../types.js';

export interface EntityCache
  extends Map<
    string,
    Map<
      string,
      WeakMap<EntityInterface, WeakDependencyMap<EntityPath, object, any>>
    >
  > {}

export type EndpointsCache = WeakDependencyMap<EntityPath, object, any>;

export type IndexPath = [key: string, field: string, value: string];
export type EntitySchemaPath = [key: string] | [key: string, pk: string];
export type QueryPath = IndexPath | EntitySchemaPath;
