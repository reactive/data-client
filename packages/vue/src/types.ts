import type { MaybeRefOrGetter } from 'vue';

/** Maps each parameter to accept raw value, Ref, ComputedRef, or getter */
export type MaybeRefsOrGetters<T extends readonly any[]> = {
  readonly [K in keyof T]: MaybeRefOrGetter<T[K]>;
};

/** Maps each parameter to accept raw value, Ref, ComputedRef, or getter, with nullable support */
export type MaybeRefsOrGettersNullable<T extends readonly any[]> = {
  readonly [K in keyof T]: MaybeRefOrGetter<T[K] | null>;
};
