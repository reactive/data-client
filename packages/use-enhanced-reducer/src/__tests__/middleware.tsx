import { renderHook, act } from '../../../test';
import { MiddlewareAPI } from '../types';
import useEnhancedReducer from '../useEnhancedReducer';

function ignoreError(e: Event) {
  e.preventDefault();
}
beforeEach(() => {
  if (typeof addEventListener === 'function')
    addEventListener('error', ignoreError);
});
afterEach(() => {
  if (typeof removeEventListener === 'function')
    removeEventListener('error', ignoreError);
});

describe('createEnhancedReducerHook', () => {
  const makeTestActionMiddleware = (test: (...args: any) => any) => () => {
    return (next: any) => (action: any) => {
      test(action);
      return next(action);
    };
  };
  const makeTestMiddleware =
    (spy: (...args: any) => any) => (methods: MiddlewareAPI) => {
      spy(methods);
      return (next: any) => (action: any) => next(action);
    };
  const dispatchingMiddleware = ({ dispatch }: MiddlewareAPI) => {
    return (next: any) => (action: any) => {
      if (action.type === 'dispatch') {
        dispatch({ ...action, type: 'nothing' });
      }
      return next(action);
    };
  };
  const makeStatefulMiddleware =
    ({
      callBefore,
      callAfter,
    }: {
      callBefore: (...args: any) => any;
      callAfter: (...args: any) => any;
    }) =>
    ({ getState }: MiddlewareAPI) => {
      return (next: any) => async (action: any) => {
        callBefore(getState());
        await next(action);
        callAfter(getState());
      };
    };

  test('runs through zero middlewares', () => {
    const { result } = renderHook(() => {
      return useEnhancedReducer(state => state, {}, []);
    });
    const [state, dispatch] = result.current;

    const action = { type: 5 };
    act(() => {
      dispatch(action);
    });
    act(() => {
      dispatch(action);
    });
  });

  test('runs through a single middleware', () => {
    const faker = jest.fn();
    const logger = makeTestActionMiddleware(faker);
    const { result } = renderHook(() => {
      return useEnhancedReducer(state => state, {}, [logger]);
    });
    const [state, dispatch] = result.current;

    expect(faker.mock.calls.length).toBe(0);
    const action = { type: 5 };
    act(() => {
      dispatch(action);
    });
    expect(faker.mock.calls.length).toBe(1);
    expect(faker.mock.calls[0][0]).toBe(action);
    act(() => {
      dispatch(action);
    });
    expect(faker.mock.calls.length).toBe(2);
  });

  it('wraps dispatch method with middleware once', () => {
    const [faker, statefaker] = [jest.fn(), jest.fn()];
    const methodspy = makeTestMiddleware(faker);
    const statespy = makeTestMiddleware(({ getState }: MiddlewareAPI) =>
      statefaker(getState()),
    );

    const { result } = renderHook(() => {
      return useEnhancedReducer(state => state, { double: 5 }, [
        methodspy,
        statespy,
      ]);
    });
    const [state, dispatch] = result.current;

    const action = { type: 5 };
    act(() => {
      dispatch(action);
      dispatch(action);
    });

    expect(faker.mock.calls.length).toEqual(1);

    expect(faker.mock.calls[0][0]).toHaveProperty('getState');
    expect(faker.mock.calls[0][0]).toHaveProperty('dispatch');

    expect(statefaker.mock.calls[0][0]).toEqual({ double: 5 });
  });

  test('reducer to work properly', () => {
    const logger = makeTestActionMiddleware(() => {});

    const { result } = renderHook(() => {
      return useEnhancedReducer(
        (state, action) => ({ ...state, omlet: action.payload }),
        { eggs: 'bacon' },
        [logger],
      );
    });
    let [state, dispatch] = result.current;
    act(() => {
      dispatch({ payload: 5 });
    });
    [state, dispatch] = result.current;
    expect(state).toEqual({
      eggs: 'bacon',
      omlet: 5,
    });
    act(() => {
      dispatch({ payload: 'chicken' });
    });
    [state, dispatch] = result.current;
    expect(state).toEqual({
      eggs: 'bacon',
      omlet: 'chicken',
    });
  });

  test('should work with middlewares that call dispatch', () => {
    const faker = jest.fn();
    const logger = makeTestActionMiddleware(faker);

    const { result } = renderHook(() => {
      return useEnhancedReducer(state => state, {}, [
        logger,
        dispatchingMiddleware,
      ]);
    });
    const [state, dispatch] = result.current;
    expect(faker.mock.calls.length).toBe(0);
    let action: any = { type: 'hi' };
    act(() => {
      dispatch(action);
    });
    expect(faker.mock.calls.length).toBe(1);
    expect(faker.mock.calls[0][0]).toBe(action);
    action = { type: 'dispatch', payload: 5 };
    act(() => {
      dispatch(action);
    });
    expect(faker.mock.calls.length).toBe(3);
    expect(faker.mock.calls[2][0]).toEqual({ type: 'nothing', payload: 5 });
  });

  test('should work with middlewares that getState()', async () => {
    const callBefore = jest.fn();
    const callAfter = jest.fn();
    const logger = makeStatefulMiddleware({ callBefore, callAfter });

    const reducer = (state: { counter: number }) => ({
      ...state,
      counter: state.counter + 1,
    });
    const { result, waitFor } = renderHook(() => {
      return useEnhancedReducer(reducer, { counter: 0 }, [logger]);
    });
    let [state, dispatch] = result.current;
    expect(callBefore.mock.calls.length).toBe(0);
    let action: any = { type: 'hi' };
    act(() => {
      dispatch(action);
    });

    [state, dispatch] = result.current;
    expect(callBefore.mock.calls.length).toBe(1);
    await waitFor(() => {
      expect(callAfter.mock.calls.length).toBe(1);
    });
    expect(callBefore.mock.calls[0][0]).toEqual({ counter: 0 });
    expect(callAfter.mock.calls[0][0]).toEqual({ counter: 1 });
    expect(callAfter.mock.calls[0][0]).toEqual(state);
    action = { type: 'dispatch', payload: 5 };
    act(() => {
      dispatch(action);
    });
    expect(callBefore.mock.calls.length).toBe(2);
    await waitFor(() => {
      expect(callAfter.mock.calls.length).toBe(2);
    });
    expect(callBefore.mock.calls[1][0]).toEqual({ counter: 1 });
    expect(callAfter.mock.calls[1][0]).toEqual({ counter: 2 });
  });

  it('warns when dispatching during middleware setup', () => {
    const consoleErrorSpy = jest
      .spyOn(console, 'error')
      .mockImplementation(() => {});
    try {
      function dispatchingMiddleware({
        dispatch,
      }: {
        dispatch: (...args: any) => any;
      }) {
        dispatch({ type: 'dispatch', payload: 5 });
        return (next: (...args: any) => any) => (action: any) => next(action);
      }
      const { result } = renderHook(() => {
        useEnhancedReducer(state => state, {}, [dispatchingMiddleware]);
      });
      expect(result.error).toBeDefined();
      expect(result.error?.message).toMatchSnapshot();
    } finally {
      consoleErrorSpy.mockRestore();
    }
  });
});
