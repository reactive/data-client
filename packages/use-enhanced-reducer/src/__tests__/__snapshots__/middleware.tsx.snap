// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`createEnhancedReducerHook warns when dispatching during middleware setup 1`] = `"Dispatching while constructing your middleware is not allowed. Other middleware would not be applied to this dispatch."`;
